test/Parse/matching_patterns.swift

```diff
@@ -289,22 +289,44 @@ case (let (_, _, _)) + 1:
   ()
 }
 
-// FIXME: We don't currently allow subpatterns for "isa" patterns that
-// require interesting conditional downcasts.
-class Base { }
-class Derived : Base { }
+// "isa" patterns.
 
+// https://github.com/apple/swift/issues/56139
+// Allow subpatterns for "isa" patterns that require conditional
+// collection downcasts
+do {
+  class Base { }
+  class Derived : Base { }
 
-switch [Derived(), Derived(), Base()] {
-case let ds as [Derived]: // expected-error{{collection downcast in cast pattern is not implemented; use an explicit downcast to '[Derived]' instead}}
-  ()
-case is [Derived]: // expected-error{{collection downcast in cast pattern is not implemented; use an explicit downcast to '[Derived]' instead}}
-  ()
+  var arr: [Base]
 
-default:
-  ()
-}
+  if case let _ as [Derived] = arr {}
+
+  guard case let _ as [Derived] = arr else {}
+
+  while case let _ as [Derived] = arr {}
+
+  switch arr {
+  case let ds as [Derived]:
+    // expected-warning@-1 {{immutable value 'ds' was never used; consider replacing with '_' or removing it}}
+    ()
+  case is [Derived]:
+    ()
```

Comment:
Optional: These case elements could be compact one-liners. Also, how about a conventional `break` instead of a paren expression?